File: Chapter 03/C#/Section_3_3_5_DIP/After/Coach.cs
using System.Collections.Generic;

namespace Chapter03.DIP.After;

/// <summary>
/// DIP SOLUTION: Constructor Injection and Loose Coupling.
/// 
/// ARCHITECTURE NOTE: The Coach is no longer responsible for building its own 
/// team. The 'new' keyword has been completely removed! 
/// 
/// Instead, the Coach asks for its dependencies through its constructor. This 
/// is called "Dependency Injection." Because the Coach only asks for a list 
/// of 'IPlayer' abstractions, we can hand it literally any combination of players. 
/// 
/// This makes the Coach class infinitely extensible and incredibly easy to unit 
/// test (we can just pass in "Mock" players during testing).
/// </summary>
public class Coach
{
    // The Coach depends strictly on an abstraction!
    private readonly List<IPlayer> _team;

    /// <summary>
    /// Dependencies are provided from the outside (Constructor Injection).
    /// </summary>
    public Coach(List<IPlayer> players)
    {
        _team = players;
    }

    public void ExecuteGamePlan()
    {
        foreach (var player in _team)
        {
            player.PerformAction();
        }
    }
}

File: Chapter 03/JS/section_3_3_5_dip/after/coach.js
/**
 * DIP SOLUTION: Constructor Injection and Loose Coupling.
 * * ARCHITECTURE NOTE: The Coach is no longer responsible for building its own 
 * team. The 'new' keyword and hardcoded 'requires' have been completely removed!
 * * Instead, the Coach receives its dependencies as an argument. This is 
 * "Dependency Injection." Because the Coach only expects an array of objects 
 * with a .performAction() method, we can hand it any combination of players.
 * * This makes the Coach class infinitely extensible and trivial to unit test 
 * using mock player objects.
 */
class Coach {
    /**
     * Dependencies are provided from the outside.
     * @param {Array} players - A list of player objects.
     */
    constructor(players) {
        this.team = players;
    }

    executeGamePlan() {
        // The Coach now relies on a shared "contract" (performAction)
        // rather than specific concrete methods like .attack() or .tackle().
        for (const player of this.team) {
            player.performAction();
        }
    }
}

module.exports = Coach;